Create at most one PrintStream per log file.

hwellmann opened the pull request:
This avoids corrupted log files.

Fixes #652.

rhuss commented on src/main/java/io/fabric8/maven/docker/log/LogCallbackFactory.java:
```diff
@@ -0,0 +1,39 @@
+        return ps;
+    }
+
+    public static void closeLogs() {
```
This method should be `synchronized`, too

rhuss commented on src/main/java/io/fabric8/maven/docker/log/LogDispatcher.java:
```diff
@@ -37,15 +40,16 @@ public LogDispatcher(DockerAccess dockerAccess) {
 
     public synchronized void trackContainerLog(String containerId, LogOutputSpec spec) throws FileNotFoundException {
-        LogGetHandle handle = dockerAccess.getLogAsync(containerId, new DefaultLogCallback(spec));
+        LogGetHandle handle = dockerAccess.getLogAsync(containerId, createLogCallback(spec));
```
I'm not a a fan of static imports as they dilute the real locations and I always search first in the current class when doing review. Could we refer to the static method via its class (`LogCallbackFactory.createLogBack()`) ? Even saves some lines.

rhuss commented on src/main/java/io/fabric8/maven/docker/access/hc/DockerAccessWithHcClient.java:
```diff
@@ -145,7 +146,7 @@ public void startExecContainer(String containerId, LogOutputSpec outputSpec) thr
 
     private ResponseHandler<Object> createExecResponseHandler(LogOutputSpec outputSpec) throws FileNotFoundException {
-        final LogCallback callback = new DefaultLogCallback(outputSpec);
+        final LogCallback callback = createLogCallback(outputSpec);
```
Who is closing this DefaultLogCallback() ?

rhuss commented on src/main/java/io/fabric8/maven/docker/log/LogDispatcher.java:
```diff
@@ -37,15 +40,16 @@ public LogDispatcher(DockerAccess dockerAccess) {
     }
 
     public synchronized void untrackAllContainerLogs() {
+        closeLogs();
```
This will close all streams from all LogDispatcher instances running. Not sure why it could be more than one, but at least we should avoid the mismatch with the LogCallbackFactory storing the stream in a static Map vs. using object instances for the LogDispatcher here. I would prefer to have LogCallbackFactory to be an object, too.

What's about the following idea: We simply allow `DefaultLogCallback` to reuse an existing printstream (i.e. we do the caching on a static field in DefaultLogCallback along with a usage count) and add a `close()` method to the callback so that the user can signal that he is done with logging ? In `close()` an the associated printstream's usage count will be decreased and if 0 then closed.

I think this kind of resource management makes more sense, also for potential future use cases. If you don't have the cycles to implement this, no problem, I could take over (as this is indeed an important fix to have). Also, a unit tests for `DefaultLogCallback` would be awesome then.

hwellmann commented:
Thanks for your feedback. I'll get back to this tomorrow...

rhuss commented:
Awesome,thx unfortunately I just catched an ugly flu, and staying in bed
currently. I will continue here asap. Sorry ...

Harald Wellmann <notifications@github.com> schrieb am Di., 20. Dez. 2016,
13:44:

> Thanks for your feedback. I'll get back to this tomorrow...
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/fabric8io/docker-maven-plugin/pull/661#issuecomment-268234427>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGDCF5LZDD0ujqAvS_aVoM9v4dxa36oks5rJ82pgaJpZM4LP7z2>
> .
>

hwellmann commented:
Reworked `DefaultLogCallback` with reference counting and added a test class. The `LogCallbackFactory` is gone.

Note: To test this in the context of my real-life project, I had to revert 7afec26aaaa9ebe39065a8daa8a96db42945a22e, see #660. 

Master appears to be broken, I'm seeing the same issue both on MacOS and Linux.

rhuss commented on src/main/java/io/fabric8/maven/docker/log/SharedPrintStream.java:
```diff
@@ -0,0 +1,35 @@
+        return printStream;
+    }
+
+    public void setPrintStream(PrintStream printStream) {
```
Doesnt look like that this method is ever used. Also, removing this set-method makes the class immutable with respect to the printstream which is a good thing :)

rhuss commented on src/main/java/io/fabric8/maven/docker/access/log/LogRequestor.java:
```diff
@@ -91,6 +91,8 @@ public void fetchLogs() {
         } catch (IOException exp) {
             callback.error(exp.getMessage());
+        } finally {
+            callback.close();
```
We should ensure, that, when the callback is closed that this any call to the `LogRequestors` method is invalid (by throwing an exception).

Its also not so nice that the responsibilities for creating the callback (== caller of constructor of `LogRequestor`) and the one closing the callback (== the `LogRequestor`) are different objects. It would be much better when the `LogRequestor` has the complete control over allocation and closing of the print stream.

I suggest therefor to introduce a an `LogCallback open()` which allocates the stream instead of automatically allocating it in the constructor. Each method which can close a stream then would have to open it before.

rhuss commented on src/main/java/io/fabric8/maven/docker/access/log/LogRequestor.java:
```diff
@@ -224,6 +227,7 @@ public void finish() {
                 }
             }
         }
+        callback.close();
```
finish() is called in case of an exception but the `callback.close()` is also included in the finally clauses of the top level methods. This leads to closing a callback() twice which is probably not intended.

rhuss commented on src/test/java/io/fabric8/maven/docker/log/DefaultLogCallbackTest.java:
```diff
@@ -0,0 +1,173 @@
+        callback2.log(1, ts, "line 4");
+        callback2.close();
+
+        List<String> lines = Files.readAllLines(path);
```
Unfortunately we can't use Java 8 and are still stuck to Java 7 for now (this might change in the future). So we need some other way to read all lines here.

rhuss commented:
@hwellmann I made some changes and would like to push it to this PR. Could you please grant me push permissions ? This is described [here](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)

rhuss commented:
Sorry, my faul. Mispelled the target branch for the push.

rhuss commented:
[merge]
